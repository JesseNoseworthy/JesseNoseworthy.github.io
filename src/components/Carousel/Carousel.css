@import '../../styles/variables.css';
@import '../../styles/mixins.css';
@import './slick/slick.css';

:local(.Carousel) {
  &:not(:local(.isVerticalCarousel)) {
    :global {
      .slick-dots {
        bottom: 35px;
        li {
          margin: 0;
          width: $carouselDotSize;
          height: $carouselDotSize;
          line-height: $carouselDotSize;
        }
        button {
          width: $carouselDotSize;
          height: $carouselDotSize;
          line-height: $carouselDotSize;
          &:before {
            width: 5px;
            height: 5px;
            position: absolute;
            top: 0;
            left: 0;
            border-radius: 50%;
          }
        }
      }
      .slick-prev,
      .slick-next {
        height: 42px;
        width: 42px;
        z-index: 1;
        svg g {
          stroke: $red;
        }
        &:before,
        &:after {
          content: ''
        }
      }
      .slick-prev {
        left: 19px;
      }
      .slick-next {
        right: 19px;
      }
    }
  }
  &:not(:local(.CarouselFullWidthMobile)) {
    :global {
      @media (--mobile) {
        .slick-slide {
          width: 95vw !important;
          margin-right: 4px;
        }
      }
    }
  }
  :global {
    .slick-slide {
      border: none !important;
    }
    :not(.slick-active) {
      button {
        &:before {
          color: $darkGrey !important;
          opacity: 0.5 !important;
        }
      }
    }
    .slick-active {
      button {
        &:before {
          color: $red !important;
          opacity: 1 !important;
        }
      }
    }
  }
  &:global(.slick-slider) {
    margin-bottom: 0 !important;
  }
}

:local(.CarouselFullWidthMobile) {
  :global {
    @media (--mobile) {
      .slick-slide {
        width: 100vw !important;
      }
    }
  }
}

:local(.isVerticalCarousel) {
  :global {
    .slick-dots {
      bottom: 50%;
      text-align: left;
      display: flex !important;
      flex-direction: column;
    }
  }
}