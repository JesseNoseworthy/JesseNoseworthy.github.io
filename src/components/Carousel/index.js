import React, { Component } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import styles from './Carousel.css';
import Slider from 'react-slick';
import Button from 'components/Button';
import CanonIcon from 'components/CanonIcon';
import { isMobile } from '../../utils';

const ArrowButton = ({
  icon,
  currentSlide,
  slideCount,
  ...others
}) => (
  <Button {...others}>
    <CanonIcon icon={icon} />
  </Button>
);

class Carousel extends Component {
  constructor(props) {
    super(props);
    
    this.state = {
      touchStartXAxis: null,
      touchMoveXAxis: null,
      validDragThreshold: false
    }

    this.onTouchStart = this.onTouchStart.bind(this);
    this.onTouchMove = this.onTouchMove.bind(this);
  }

  getAutoplaySettings(isAutoplaying) {
    if (isAutoplaying) {
      return {
        autoplay: true,
        autoplaySpeed: 5000
      }
    }
  }

  componentDidMount() {
    window.addEventListener('touchstart', this.onTouchStart);
    window.addEventListener('touchmove', this.onTouchMove);
  }

  componentWillUnmount() {
    window.removeEventListener('touchstart', this.onTouchStart);
    window.removeEventListener('touchmove', this.onTouchMove);
  }

  onTouchStart(e) {
    this.setState({
      touchStartXAxis: e.changedTouches[0].pageX,
    })
  }

  onTouchMove(e) {
    this.setState({
      touchMoveXAxis: e.changedTouches[0].pageX,
      validDragThreshold: e.changedTouches[0].pageX > this.state.touchStartXAxis
    })
  }

  render() {
    const {
      children,
      isVertical,
      isFullWidthMobile,
      className,
      isAutoplaying,
      ...additionalOptions
    } = this.props;

    const settings = {
      infinite: true,
      adaptiveHeight: true,
      dots: true,
      arrows: !isVertical,
      prevArrow: <ArrowButton icon='PreviousArrow' />,
      nextArrow: <ArrowButton icon='NextArrow' />,
      vertical: isVertical && !isMobile(),
      verticalSwiping: isVertical && !isMobile(),
      responsive: [
        {
          breakpoint: 767,
          settings: {
            arrows: false,
            variableWidth: true,
            slidesToShow: 1,
            swipe: this.state.validDragThreshold
          }
        }
      ],
      ...this.getAutoplaySettings(isAutoplaying),
      ...additionalOptions
    }

    return children.length > 1 ? (
      <Slider
        className={classNames(
          styles.Carousel,
          className,
          { [styles.isVerticalCarousel]: isVertical && !isMobile() },
          { [styles.CarouselFullWidthMobile]: isFullWidthMobile }
        )}
        ref='carouselSlider'
        {...settings}>
        {children}
      </Slider>
      ) : (
      <div>
        {children}
      </div>
    )
  }
}

Carousel.propTypes = {
  className: PropTypes.string,
  additionalOptions: PropTypes.object,
  isVertical: PropTypes.bool,
  isFullWidthMobile: PropTypes.bool
};

Carousel.defaultProps = {
  isVertical: false,
  isAutoplaying: true,
  isFullWidthMobile: true
};

export default Carousel;