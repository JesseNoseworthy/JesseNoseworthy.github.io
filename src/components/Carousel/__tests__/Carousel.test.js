import React from 'react';
import renderer from 'react-test-renderer';
import Carousel from '../index';

describe('Carousel', () => {
  describe('Snapshot tests', () => {
    it('It renders a simple div with one child', () => {
      const component = renderer.create(
        <Carousel>
          <img src='image src' alt='text' />
        </Carousel>
      );

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('It accepts additional options', () => {

      const additionalOptions = {
        infinite: true,
        adaptiveHeight: true
      }

      const component = renderer.create(
        <Carousel {...additionalOptions}>
          <img src='image src' alt='text' />
        </Carousel>
      );

      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
