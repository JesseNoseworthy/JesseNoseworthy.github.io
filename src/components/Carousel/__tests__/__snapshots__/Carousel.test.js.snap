exports[`Carousel Snapshot tests It accepts additional options 1`] = `
<div>
  <img
    alt="text"
    src="image src" />
</div>
`;

exports[`Carousel Snapshot tests It renders a simple div with one child 1`] = `
<div>
  <img
    alt="text"
    src="image src" />
</div>
`;
